### PR TITLE
feat: Table support `scrollTo` `reference`

### LINKED ISSUE #44595: Can't solve a problem with antd.Table
### What problem does this feature solve?

Retain the scroll position at the right place while continuously adding new row on top of the ant.Table.
I think I need some new feature to handle this situation.

Problem/question is from real life project:
I have posted a question in stackoverflow about antd.Table. Still couldn't get satisfying answer or solution.
So I started to think may be for now there is no solution for my problem using ant.Table.
So may be developers could have add some new features to solve this kind of problems
Link to stackoverflow question: https://stackoverflow.com/questions/77007923/how-can-i-retain-scroll-position-using-antd-table-in-react

### What does the proposed API look like?

Code that didn't give desired result: https://codesandbox.io/s/practical-cdn-l5v9ny?file=/src/App.js
Another code that worked with self defined <ItemList /> component: https://codesandbox.io/s/exciting-brown-xw9xpt?file=/src/App.js  
Wait until rows/items overflow the screen. Then scroll downward to see difference between codes in these links

### LINKED ISSUE #45234: Table设置虚拟滚动后需要支持滚动到第一行
### What problem does this feature solve?

使用表格虚拟滚动 Form.List做可编辑表格时，新增一行时需要自动跳到第一行或最后一行让用户填写新内容

### What does the proposed API look like?

类似tree虚拟滚动支持跳转的API

### PR DESCRIPTION
**Pull request by zombieJ**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 💡 Background and solution

ref https://github.com/ant-design/ant-design/discussions/45242
resolve #44595
resolve #45234

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |   🚨 Refactor Table `ref` to support `scrollTo` to scroll to target `key` or `index` or `top`.        |
| 🇨🇳 Chinese |     🚨 重构 Table `ref` 支持 `scrollTo` 以滚动到目标 `key` 或 `index` 或 `top`      |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at 09b71fd</samp>

Added a `reference` prop to the `Table` component to allow accessing the table dom element and scrolling methods. Updated the documentation, tests, and dependencies accordingly. Refactored some internal components and types of the `Table` component.

### <samp>🤖 Generated by Copilot at 09b71fd</samp>

*  Add `reference` prop to `Table` component to support getting the native dom element and other methods from the table instance ([link](https://github.com/ant-design/ant-design/pull/45245/files?diff=unified&w=0#diff-a5896d4aa0af160e25dae36aec024d58f1f1769814b1777acedf581c53140616R151), [link](https://github.com/ant-design/ant-design/pull/45245/files?diff=unified&w=0#diff-a5896d4aa0af160e25dae36aec024d58f1f1769814b1777acedf581c53140616R227-R237), [link](https://github.com/ant-design/ant-design/pull/45245/files?diff=unified&w=0#diff-a5896d4aa0af160e25dae36aec024d58f1f1769814b1777acedf581c53140616R593), [link](https://github.com/ant-design/ant-design/pull/45245/files?diff=unified&w=0#diff-9e016ea02fe7dd0ea42d8774d29d45e1a10b8dd108de9c89d50c828c74b993b8R14-R15), [link](https://github.com/ant-design/ant-design/pull/45245/files?diff=unified&w=0#diff-1472817fa7245b7d992a6a03b35192c613278a6248370d54eb2a8b0e62814ed1L3-R4), [link](https://github.com/ant-design/ant-design/pull/45245/files?diff=unified&w=0#diff-1472817fa7245b7d992a6a03b35192c613278a6248370d54eb2a8b0e62814ed1R405-R411), [link](https://github.com/ant-design/ant-design/pull/45245/files?diff=unified&w=0#diff-8d3a4194d516aa3322b22843bc2a379cdb0ad134c5e22899e6468d16c85d85c9R128), [link](https://github.com/ant-design/ant-design/pull/45245/files?diff=unified&w=0#diff-926ab5191876ea212b18b01284be99615b47d3d569e1c6e0c410b944c35431d2R6), [link](https://github.com/ant-design/ant-design/pull/45245/files?diff=unified&w=0#diff-1d555de1f30d5e46ff95e81b3d10b27baf15321e455557e7d3a55ca09ff61209R129), [link](https://github.com/ant-design/ant-design/pull/45245/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L150-R150))
* Move `EXPAND_COLUMN` and `Summary` from `Table.tsx` to `InternalTable.tsx` as they are only used in the latter file ([link](https://github.com/ant-design/ant-design/pull/45245/files?diff=unified&w=0#diff-a5896d4aa0af160e25dae36aec024d58f1f1769814b1777acedf581c53140616L3-R5), [link](https://github.com/ant-design/ant-design/pull/45245/files?diff=unified&w=0#diff-9e016ea02fe7dd0ea42d8774d29d45e1a10b8dd108de9c89d50c828c74b993b8L1-R6))
* Remove unnecessary `prettier-ignore` comment and extra `|` symbol from `index.zh-CN.md` file ([link](https://github.com/ant-design/ant-design/pull/45245/files?diff=unified&w=0#diff-1d555de1f30d5e46ff95e81b3d10b27baf15321e455557e7d3a55ca09ff61209L173), [link](https://github.com/ant-design/ant-design/pull/45245/files?diff=unified&w=0#diff-1d555de1f30d5e46ff95e81b3d10b27baf15321e455557e7d3a55ca09ff61209L208-R208))

### CODE REVIEW

**Comment by afc163** [comment]
关联一下 issue

**Review by kiner-tang — APPROVED** [review]

File: components/table/__tests__/Table.test.tsx [hunk 1]
```diff
@@ -401,4 +402,11 @@ describe('Table', () => {
     await waitFakeTimer();
     expect(container.querySelector('.ant-dropdown')).toBeTruthy();
   });
+
+  it('support reference', () => {
+    const tblRef = React.createRef<Reference>();
+    const { container } = render(<Table reference={tblRef} />);
```
Comment (afc163) [inline components/table/__tests__/Table.test.tsx:408]: 为啥不用 ref？

File: components/table/InternalTable.tsx [hunk 2]
```diff
@@ -1,7 +1,8 @@
 import * as React from 'react';
 import classNames from 'classnames';
-import { INTERNAL_HOOKS, type TableProps as RcTableProps } from 'rc-table';
+import { INTERNAL_HOOKS, type TableProps as RcTableProps, type Reference } from 'rc-table';
```
Comment (crazyair) [inline components/table/InternalTable.tsx:3]:
```suggestion
import { INTERNAL_HOOKS, type TableProps as RcTableProps, type Reference as RcReference } from 'rc-table';
```

File: components/table/InternalTable.tsx [hunk 3]
```diff
@@ -222,6 +224,17 @@ const InternalTable = <RecordType extends AnyObject = AnyObject>(
...
 
+  // ============================= Refs =============================
+  const rootRef = React.useRef<HTMLDivElement>(null);
+  const tblRef = React.useRef<Reference>(null);
+
+  const mergedRootRef = useComposeRef(rootRef, ref);
+
+  React.useImperativeHandle(reference, () => ({
+    ...tblRef.current!,
+    nativeElement: rootRef.current!,
```
[the remark this answers is not in the document]
Comment (zombieJ) [reply components/table/InternalTable.tsx:239]: 不会的，`useImperativeHandle` 会确保执行在 ref 注入后。

[the remark this answers is not in the document]
Comment (crazyair) [reply components/table/InternalTable.tsx:239]: ok

File: components/table/__tests__/Table.test.tsx (hunk 1 above)
Comment (zombieJ) [reply components/table/__tests__/Table.test.tsx:408]: `ref` 指向的是原生 dom，原生 dom 有自己的 `scrollTo` 方法。最好的实现应该是像  mobile 一样，`ref` 里返回一个包含 `nativeElement` 的对象。但是对 antd 来说，这个就已经不能改了。不同组件的 `ref` 给的都不一样。所以统一 `reference` 返回统一的对象。

File: components/table/index.en-US.md [hunk 4]
```diff
@@ -125,6 +125,7 @@ Common props ref：[Common props](/docs/react/common-props)
 | loading | Loading status of table | boolean \| [Spin Props](/components/spin/#api) | false |  |
 | locale | The i18n text including filter, sort, empty text, etc | object | [Default Value](https://github.com/ant-design/ant-design/blob/6dae4a7e18ad1ba193aedd5ab6867e1d823e2aa4/components/locale/en_US.tsx#L19-L37) |  |
 | pagination | Config of pagination. You can ref table pagination [config](#pagination) or full [`pagination`](/components/pagination/) document, hide it by setting it to `false` | object \| `false` | - |  |
+| reference | Similar as `ref`. Support get native dom or other function | { nativeElement: HTMLDivElement, scrollTo: (config: { index?: number, key?: React.Key, top?: number }) => void } | - | 5.11.0 |
```
Comment (crazyair) [inline components/table/index.en-US.md:128]: top 需要支持 `px` `vh`？

Comment (crazyair) [reply components/table/index.en-US.md:128]: key 要不要说明对应的是 rowKey

File: components/table/__tests__/Table.test.tsx (hunk 1 above)
Comment (afc163) [reply components/table/__tests__/Table.test.tsx:408]: 类似 form 的 scrollToField，这个功能改成另一个名字 scrollToRow 和 scrollToColumn 如何呢？

Comment (afc163) [reply components/table/__tests__/Table.test.tsx:408]: reference 这个一旦开了就收不回来了。

Comment (zombieJ) [reply components/table/__tests__/Table.test.tsx:408]: `scrollTo` 这个是没问题的，Tree 里也是这个 API。form 是有个 Form instance，才有的 `scrollToField`。如果按照这个逻辑设计就太复杂了：
```tsx
const table = Table.useTable();

<Table table={table} />;
```

`reference` 这个 API 设计就是为了不做 breaking change 去收口，以后所有的 `ref` 都走这个。

Comment (crazyair) [reply components/table/__tests__/Table.test.tsx:408]: 这样也挺好 `<Form form={form} />` `<Table table={table} />`

File: components/table/index.zh-CN.md [hunk 5]
```diff
@@ -143,6 +143,13 @@ const columns = [
 | onRow | 设置行属性 | function(record, index) | - |  |
 | virtual | 支持虚拟列表 | boolean | - | 5.9.0 |
 
+### Table ref
+
+| Property | Description | Type | Version |
```
Comment (yoyo837) [inline components/table/index.zh-CN.md:148]: 这里是中文, 而且怎么只有4列, 应该5列吧

File: components/table/index.en-US.md (hunk 4 above)
Comment (zombieJ) [reply components/table/index.en-US.md]: `top` 和原生的一样，只支持 `px`。`key` 这个可以补充说明一下。

File: components/table/index.zh-CN.md (hunk 5 above)
Comment (zombieJ) [reply components/table/index.zh-CN.md:148]: OK & 这个是 ref API，没有默认值。

File: components/table/demo/virtual-list.tsx [hunk 6]
```diff
@@ -114,6 +114,7 @@ const App = () => {
   const [empty, setEmpty] = React.useState(false);
   const [count, setCount] = React.useState(10000);
 
+  const tblRef: Parameters<typeof Table>[0]['ref'] = React.useRef(null);
```
Comment (crazyair) [inline components/table/demo/virtual-list.tsx:117]: 类型不能放到 `useRef<xxx>` 这里吗

Comment (zombieJ) [reply components/table/demo/virtual-list.tsx:117]: 那就要从 `import { type TableRef } from 'antd/lib/table';` 里调用了，DEMO 里一般搞推导。这样 Codesandbox 里引入也方便。

Comment (crazyair) [reply components/table/demo/virtual-list.tsx:117]: 也是推荐用户编写风格的，用推荐写法比较好，上面这种可以在 test 里随便写

File: components/table/index.en-US.md (hunk 4 above)
Comment (crazyair) [reply components/table/index.en-US.md]: 支持 `px`，类型要加下 `string`？

**Comment by qingfengmy** [comment]
有文档吗，怎么拿到rc-vitual-list的ref，我需要用类似scrollToItem的方法去滚动到某个为止

**Comment by qingfengmy** [comment]
图片


这个是要到5.11.0版本才支持吗？现在antd版本才5.10.3，没有这个功能是吗？

**Comment by xv-jiaqi** [comment]
这个 feature 是违反发版约定的破坏性升级，让原使用 Table 组件 ref 属性用户在小版本升级（5.8.6 -> 5.11.0）后直接报错。

image

**Pull request merged**